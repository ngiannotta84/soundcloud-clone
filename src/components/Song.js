import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { music } from "../media/icons";
import "../styles/song.css";

const Song = ({
  artistName,
  albumName,
  albumArt,
  songName,
  songAudio,
  handleSetPlaylist,
}) => {
  const handleClick = (addNext = false) => {
    const data = {
      image: albumArt,
      artistName,
      albumName,
      songName,
      audio: songAudio,
    };
    handleSetPlaylist(data, addNext);
  };

  return (
    <div className="song">
      <div className="song__info-container">
        <img
          src={albumArt || music}
          alt={`${albumName} cover art`}
          className="song__cover-art"
        />
        <div className="song__info">
          <h2 className="song__artist-name">
            <Link
              to={`/profile/${artistName}`}
              className="song__artist-name__link"
            >
              {artistName}
            </Link>
          </h2>
          <h3 className="song__song-name">{songName}</h3>
        </div>
      </div>
      <div className="song__buttons">
        <button
          type="button"
          className="song__button"
          onClick={() => handleClick()}
        >
          Add to queue
        </button>
        <button
          type="button"
          className="song__button"
          onClick={() => handleClick(true)}
        >
          Play next
        </button>
      </div>
    </div>
  );
};

Song.propTypes = {
  artistName: PropTypes.string.isRequired,
  albumName: PropTypes.string.isRequired,
  albumArt: PropTypes.string.isRequired,
  songName: PropTypes.string.isRequired,
  songAudio: PropTypes.string.isRequired,
  handleSetPlaylist: PropTypes.func.isRequired,
};

export default Song;
