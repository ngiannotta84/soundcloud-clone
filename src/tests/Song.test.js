import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Song from "../components/Song";

describe("Song", () => {
  const validProps = {
    artistName: "artist",
    albumName: "album",
    albumArt: "alburl",
    songName: "song",
    songAudio: "songurl",
    handleSetPlaylist: jest.fn(),
  };

  test("snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Song {...validProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(
        <Router>
          <Song {...validProps} />
        </Router>
      );
    });

    test("renders correctly", () => {
      expect(screen.getByRole("img")).toHaveAttribute(
        "alt",
        `${validProps.albumName} cover art`
      );
      expect(screen.getByText(validProps.artistName)).toHaveAttribute(
        "href",
        `/profile/${validProps.artistName}`
      );
      expect(screen.getByText(validProps.songName)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByText(/add to queue/i)).toBeInstanceOf(
        HTMLButtonElement
      );
      expect(screen.getByText(/play next/i)).toBeInstanceOf(HTMLButtonElement);
    });

    test("add to queue button", () => {
      const button = screen.getByText(/add to queue/i);
      const data = {
        image: validProps.albumArt,
        artistName: validProps.artistName,
        albumName: validProps.albumName,
        songName: validProps.songName,
        audio: validProps.songAudio,
      };

      fireEvent.click(button);

      expect(validProps.handleSetPlaylist).toBeCalledWith(data, false);
    });

    test("play next button", () => {
      const button = screen.getAllByText(/play next/i)[0];
      const data = {
        image: validProps.albumArt,
        artistName: validProps.artistName,
        albumName: validProps.albumName,
        songName: validProps.songName,
        audio: validProps.songAudio,
      };

      fireEvent.click(button);

      expect(validProps.handleSetPlaylist).toBeCalledWith(data, true);
    });
  });
});
