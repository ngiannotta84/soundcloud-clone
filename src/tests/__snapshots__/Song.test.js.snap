// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song snapshot 1`] = `
<DocumentFragment>
  <div
    class="song"
  >
    <div
      class="song__info-container"
    >
      <img
        alt="album cover art"
        class="song__cover-art"
        src="alburl"
      />
      <div>
        <h2
          class="song__artist-name"
        >
          <a
            href="/profile/artist"
          >
            artist
          </a>
        </h2>
        <h3
          class="song__album-name"
        >
          song
        </h3>
      </div>
    </div>
    <div>
      <button
        class="album__song__add-song"
        type="button"
      >
        Add to queue
      </button>
      <button
        class="album__song__add-song"
        type="button"
      >
        Play next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
